sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,130 @@
+
+
+## Kube-proxy slow in updating iptables when large no. of backends exist
+
+#### Issue
+
+When we’re running services worth a total of 120k endpoints (backends) on our 5k-node scalability tests, we’re noticing that kube-proxy is timing out while trying to do iptables-restore due to failing to acquire lock over iptables.
+
+#### Relevant issues/links
+
+- [#48107](https://github.com/kubernetes/kubernetes/issues/48107)
+
+#### Possible Solution(s)
+
+- Give more time for kube-proxy to update iptables by increasing its batching interval for large clusters
+- Decide to support a smaller no. of backends for kubernetes officially
+
+
+## Apiserver spending too much cpu/memory in duplicate Endpoints serializations
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our scale tests).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible ([#60067](https://github.com/kubernetes/kubernetes/pull/60067) was an attempt to do this, but it didn't seem to help too much)
+- Do better batching of EP updates in endpoints-controller: This is more like a bandaid, as the duplicate serializations would still happen, but we at least reduce the total no. of watch events (at the cost of increased latency of ep updates potentially). Also, this would not help too much with cases where pods of services are toggling between Ready and Unready states, or when pods are not created in a 'bursty' manner. Since we probably need to solve such cases too, solving the problem this way may be an unnecessary investment.
+
+
+## Endpoints write traffic is quadratic in the no. of endpoints
+
+#### Issue
+
+There were some [improvements](https://github.com/kubernetes/kubernetes/issues/47597#issuecomment-311698194) around endpoints-controller, but one issue that still remains that we couldn’t make progress on was to allow useful patching for EP objects (i.e patch contains only the endpoint entry to be changed). As a result, when one pod changes, you update the endpoints object with that and send the entire object (consisting of other already-present endpoints). So the average size of a write is O(N) and there are N writes in total, leading to O(N^2) write traffic. Also note how this can increase space consumption in etcd which stores multiple revisions of the object. The current endpoints API makes it hard to solve this problem without introducing breaking changes.
```

Comment:
This description is too verbose in my opinion - this after "So the average size" doesn't provide much value.

At the same time, important aspects are missing here.
Even if we would be able to send patches and if etcd space wouldn't be a problem, that is not the whole problem.

Because even if there is *one* reader of endpoints objects, that reader will be quadratic, because there is no way to read just deltas on the object - you read the whole object always.

So this whole section should be rephrased to reflect that (that basically proves that the only possible solution is redesigning api - nothing else can be done).
BTW - you should also mention that there are some caveats for the potential new API (like pretty much no matching is possible then, or more metadata is being send).

Reply:
You're right - I've tried to push in so many different concerns (well there indeed are so many problems with endpoints :p) without properly structuring it.
Reorganized now, PTAL.